src/cargo/util/toml/targets.rs

```diff
@@ -928,6 +932,54 @@ fn validate_unique_build_scripts(scripts: &[String]) -> CargoResult<()> {
     Ok(())
 }
 
+// Will check if `path` specified for a target is a valid source file
+fn validate_target_path_as_source_file(
+    path: &Path,
+    target_name: &str,
+    target_kind: &str,
+) -> CargoResult<()> {
+    if !path.exists() {
+        anyhow::bail!(
+            "can't find {} `{}` at path `{}`",
+            target_kind,
+            target_name,
+            path.display()
+        );
+    }
+
+    // If the path is likely a crate, then suggest setting the path to the entrypoint
+    if path.is_dir() {
+        anyhow::bail!(
+            "path `{}` for {} `{}` is a directory, but a source file was expected.\n\
+Consider setting the path to the intended entrypoint file instead: `{}/.../main.rs`",
```

Comment:
What if we talked the directory to report concrete `lib.rs` and `main.rs`s they could point to?